src/main/java/org/kohsuke/github/GitHubClient.java

```diff
@@ -632,11 +651,15 @@ private static IOException interpretApiError(IOException e,
 
     private static void logRetryConnectionError(IOException e, URL url, int retries) throws IOException {
         // There are a range of connection errors where we want to wait a moment and just automatically retry
+        long sleepTime = minRetryInterval;
+        if (maxRetryInterval > minRetryInterval) {
+            sleepTime = ThreadLocalRandom.current().nextLong(minRetryInterval, maxRetryInterval);
+        }
         LOGGER.log(INFO,
-                e.getMessage() + " while connecting to " + url + ". Sleeping " + GitHubClient.retryTimeoutMillis
+                e.getMessage() + " while connecting to " + url + ". Sleeping " + sleepTime
```

Comment:
I recommend creating a trace ID here for debug logging.  That way an admin can search debug logs and find related logs for a single retry sequence.  Here's an example

* log https://github.com/jenkinsci/scm-filter-jervis-plugin/blob/80e28289ede66fa18553b4e0ca5f518a4bd782bc/src/main/groovy/net/gleske/scmfilter/impl/trait/JervisFilterTrait.groovy#L144
* creating trace ID via sha256sum https://github.com/jenkinsci/scm-filter-jervis-plugin/blob/80e28289ede66fa18553b4e0ca5f518a4bd782bc/src/main/groovy/net/gleske/scmfilter/impl/trait/JervisFilterTrait.groovy#L245

When I enable debug logging for a class in the mentioned class it is so active in parallel that all of the logs come in out of order.  Because of that, using the `trace-` ID as a prefix to all of the logs enable me to search for a series of logs along with their retries.  It enabled me to find the maximum retry count across logs as well which helps an admin with tuning.

For exmaple, I default to retries of 30 in my class but I found in practice with GitHub it could retry up to 28 times.  Because that was so close to the max retry limit I increased the retry limit to 60 in my particular setup.

I also set the minimum time between retries to be 1000ms and the maximum to be 3000ms.  I've found GitHub requiring me to retry up to 1 minute in these scenarios because of secondary API limits.

The new secondary API limits are very aggressive at the moment.

Reply:
I realize the logging mechanisms have to change a little bit with my feedback; they're not as straightforward as making the change just in this area of code.